### PR TITLE
Use BrowserStack to run functional tests

### PR DESCRIPTION
**Pull request by madwort**
* Set up functional tests to run on BrowserStack
* Most tests pass, a couple of tests are fixed in this PR
* Still a couple of failing tests as below:

This test fails consistently, and looks like it could be fixed:

```
test_summary_totals_on_analyse_page (frontend.tests.functional.test_charts.AnalyseSummaryTotalsTest)
```

```
======================================================================
FAIL: test_summary_totals_on_analyse_page (frontend.tests.functional.test_charts.AnalyseSummaryTotalsTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/code/openprescribing/frontend/tests/functional/test_charts.py", line 106, in test_summary_totals_on_analyse_page
    self.assertEqual(element.text.strip(), value)
AssertionError: "Aug '16" != "Sep '16"
- Aug '16
+ Sep '16
```

These two tests fail intermittently with a timeout: 

```
test_map_slider (frontend.tests.functional.test_charts.MapTest) 
test_nothing_hidden_by_default (frontend.tests.functional.test_charts.SmallListTest) 
```

### CODE REVIEW

File: openprescribing/frontend/tests/functional/mock_api_server.py [hunk 1]
```diff
@@ -98,6 +98,7 @@ def do_GET(self):
         else:
             logger.error("Unhandled path %s" % self.path)
             self.send_response(requests.codes.not_found)
+            data = str(data)
```
Comment (madwort) [inline openprescribing/frontend/tests/functional/mock_api_server.py:101]: Some functional tests request `/api/1.0/spending/` from this mock api server, which isn't implemented, so it falls back to attempting to return the default `{}`, which being a `dict` doesn't support `encode('utf-8')`. If we convert to string here, our test doesn't crash.

Comment (madwort) [inline openprescribing/frontend/tests/functional/mock_api_server.py:101]: Some functional tests request `/api/1.0/spending/` from this mock api server, which isn't implemented, so it falls back to attempting to return the default `{}`, which being a dict doesn't implement `.encode('utf-8')`. If we convert to string here, our test doesn't crash.

Comment (evansd) [reply openprescribing/frontend/tests/functional/mock_api_server.py:101]: Ah you've fixed #2379!

Comment (madwort) [reply openprescribing/frontend/tests/functional/mock_api_server.py:101]: aha!

**Comment by evansd** [comment]
All those failed tests pass for me when I run these locally, but a different test which passes in CI fails. Fun times! I shall have to do further investigation.

**Comment by evansd** [comment]
Looks like this is nearly there but not quite. The test suite now passes (with 3 expected failures) but it's then running another command afterwards which fails with:
`Cannot create container for service test-production: invalid IP address in add-host: ""`

Context:
```
The TRAVIS variable is not set. Defaulting to a blank string.
The TRAVIS_JOB_NUMBER variable is not set. Defaulting to a blank string.
The TRAVIS_BUILD_NUMBER variable is not set. Defaulting to a blank string.
The TRAVIS_BRANCH variable is not set. Defaulting to a blank string.
The TRAVIS_PULL_REQUEST variable is not set. Defaulting to a blank string.
The TRAVIS_JOB_ID variable is not set. Defaulting to a blank string.
The COVERALLS_REPO_TOKEN variable is not set. Defaulting to a blank string.
The SAUCE_USERNAME variable is not set. Defaulting to a blank string.
The SAUCE_ACCESS_KEY variable is not set. Defaulting to a blank string.
The USE_SAUCELABS variable is not set. Defaulting to a blank string.
The SAUCE_HOST variable is not set. Defaulting to a blank string.
Starting openprescribing_db-test_1 ... 

Starting openprescribing_db-test_1 ... done
Pulling test-production (ebmdatalab/openprescribing-py3-base:latest)...
Cannot create container for service test-production: invalid IP address in add-host: ""
latest: Pulling from ebmdatalab/openprescribing-py3-base
Digest: sha256:80dd8366985d7c858074455a2c66e23319bf4c2721a96f3690f43c3d5106b312
Status: Downloaded newer image for ebmdatalab/openprescribing-py3-base:latest
##[error]Process completed with exit code 1.
```

@madwort I'm going to punt this back to you if that's OK as it may be something easy to fix if you've been looking at this recently. If not them, well, we tried our best...

**Comment by madwort** [comment]
ah yeah, leave that to me!

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -30,26 +30,38 @@ jobs:
...
+          local-identifier: random
+          local-logging-level: all-logs
+      - name: Checkout
+        uses: actions/checkout@v2
+      - name: Retrieve Google Cloud credentials
+        run: ./.github/scripts/decrypt_secret.sh
+        env:
+          GOOGLE_CLOUD_GITHUB_ACTIONS_PASSPHRASE: ${{ secrets.GOOGLE_CLOUD_GITHUB_ACTIONS_PASSPHRASE }}
+      - name: Functional tests
+        run: docker-compose run --service-ports test && docker-compose run --service-ports test-production
```
Comment (madwort) [inline .github/workflows/main.yml:57]: replicating what happened in Travis - running two test suites with an `&&` - `test` now pass but will separate out `test-production` (& possibly disregard)

**Comment by evansd** [comment]
There must be a German word for the sense of satisfaction you get from a green tick, mixed with the shame of knowing you got it by marking a bunch of tests xfail.

**Review by evansd — APPROVED** [review]

**Pull request merged**